TST: add tests for ?gtsvx

swallan opened the pull request:
<!-- 
Thanks for contributing a pull request! Please ensure that
your PR satisfies the checklist before submitting:
http://scipy.github.io/devdocs/dev/contributor/development_workflow.html#checklist-before-submitting-a-pr

Also, please name and describe your PR as you would write a
commit message:
https://docs.scipy.org/doc/numpy/dev/development_workflow.html#writing-the-commit-message

Note that we are a team of volunteers; we appreciate your
patience during the review process.

Again, thanks for contributing!
-->

#### Reference issue
<!--Example: Closes gh-WXYZ.-->
Adds tests for: Add wrapper for ?gtsvx #11373

#### What does this implement/fix?
<!--Please explain your changes.-->
This pull request adds tests for the ?gtsvx LAPACK routine. 
#### Additional information
<!--Any additional information you think is important.-->

swallan committed:
TST: Add initial test for lapack routine gtsvx
This commit adds tests for the lapack routine gtsvx. It parametrizes dtypes and facts. It creates a tridiagonal matrix, defines a solution x, and then generates solution b. It tests that the inputs are unmodified, outputs can be combined into the original matrix A, and that the outputs are of the correct format.

swallan committed:
TST: Add docstring, NAG manual example tests.

This commit adds the NAG manual example problems and refines the gtsvx general test.

swallan commented:
@mdhaber Here's the pull for my ?gtsvx branch. I think it should be pretty close. (edit: I had time so I added the transposal options tonight)

In addition, since the wrapper for this routine is written we can see the types of the output. `rcond`'s output type looks to be a float. I can change this test to look for that, and I assume that rcond will be then also be a float with ?ptsvx, so I can change that there, too.

Something else to consider is that to fully run this test, we'd need to use the ?gttrf routine, but it's not yet on this branch. Let me know what I should do about that!

swallan committed:
TST: Add transposal options to test

Adds transposal options and corrects typos.

mdhaber commented:
> Something else to consider is that to fully run this test, we'd need to use the ?gttrf routine, but it's not yet on this branch.

In the meantime, you can create a new branch from this one locally, merge Kai's branch with ?gttrf into that, and try running it. However, @Kai-Striega should probably wait until ?gttrf gets merged into master (and rebased, if need be) before opening a PR for this.

mdhaber commented on scipy/linalg/tests/test_lapack.py:
```diff
@@ -1705,3 +1705,183 @@ def test_getc2_gesc2():
+
+@pytest.mark.parametrize("dtype", DTYPES)
+@pytest.mark.parametrize("fact,dl_d_du_lambda",
+                         [("F", lambda dl, d, du: get_lapack_funcs('gttrf',
```
I don't see how this will work.  This lambda function returns the wrapper to gttrf, but I think you mean _execute_ gttrf on the inputs and return the results.

swallan replied:
Yes, the `(dl, d, du)` is missing. I will add that in.

mdhaber commented on scipy/linalg/tests/test_lapack.py:
```diff
@@ -1705,3 +1705,183 @@ def test_getc2_gesc2():
+    """
+    This test uses ?gtsvx to solve a random Ax=b system for each dtype.
+    It tests that the outputs define an LU matrix, that inputs are unmodified,
+    transposal options, incompatible shapes, singular matricies, and
```
```suggestion
    transposal options, incompatible shapes, singular matrices, and
```

mdhaber commented on scipy/linalg/tests/test_lapack.py:
```diff
@@ -1705,3 +1705,183 @@ def test_getc2_gesc2():
+    gtsvx = get_lapack_funcs('gtsvx', dtype=dtype)
+    # n is the length diagonal of A
+    n = 10
+    # generate facots dl, d, and du.
```
```suggestion
```

mdhaber commented on scipy/linalg/tests/test_lapack.py:
```diff
@@ -1705,3 +1705,183 @@ def test_getc2_gesc2():
+    atol = 100 * np.finfo(dtype).eps
+    # obtain routine
+    gtsvx = get_lapack_funcs('gtsvx', dtype=dtype)
+    # n is the length diagonal of A
```
```suggestion
    # Generate random tridiagonal matrix A
```

mdhaber replied:
Comments are good, but arguably some lines are simple enough that they should be lumped together.

swallan replied:
I probably got a little overzealous for some of these!

mdhaber replied:
Probably better to have too many than too few.

mdhaber commented on scipy/linalg/tests/test_lapack.py:
```diff
@@ -1705,3 +1705,183 @@ def test_getc2_gesc2():
+    dl = generate_random_dtype_array((n-1,), dtype=dtype)
+    d = generate_random_dtype_array((n,), dtype=dtype)
+    du = generate_random_dtype_array((n-1,), dtype=dtype)
+    # create A from these diagonals
```
```suggestion
```

mdhaber commented on scipy/linalg/tests/test_lapack.py:
```diff
@@ -1705,3 +1705,183 @@ def test_getc2_gesc2():
+    x = generate_random_dtype_array((n, 2), dtype=dtype)
+    # create b from x for equation Ax=b
+    b = A @ x
+    inputs_cpy = [dl.copy(), d.copy(), du.copy(), b.copy()]
```
On the other hand, for a line like _this_ a comment would be helpful since you don't use `inputs_cpy` immediately.

mdhaber commented on scipy/linalg/tests/test_lapack.py:
```diff
@@ -1705,3 +1705,183 @@ def test_getc2_gesc2():
+    gtsvx_out = gtsvx(dl, d, du, b, fact=fact, dlf=dlf_, df=df_, duf=duf_,
+                      du2=du2f_, ipiv=ipiv_)
+    dlf, df, duf, du2f, ipiv, x_soln, rcond, ferr, berr, info = gtsvx_out
+    # note that these are split into multiple lines since it was very long
```
```suggestion
```

mdhaber commented on scipy/linalg/tests/test_lapack.py:
```diff
@@ -1705,3 +1705,183 @@ def test_getc2_gesc2():
+            .format(berr.shape[0], b.shape[1]))
+
+    # test with singular matrix
+    # no need to test with fact "F" since ?pttrf already tests for these.
```
Maybe move the LU decomposition tests down here, too, right?

swallan replied:
Do you mean by the transposal options? I don't think we recreate A with the output when the input matrix is singular since it doesn't complete the factorization. (unless info == N, but it shouldn't be)

mdhaber replied:
I mean the tests that check that the factors represent an LU decomposition. You only need to check that when ?gtsvx does the factoring, not when the factors are provided.

mdhaber commented on scipy/linalg/tests/test_lapack.py:
```diff
@@ -1705,3 +1705,183 @@ def test_getc2_gesc2():
+            .format(berr.shape[0], b.shape[1]))
+
+    # test with singular matrix
+    # no need to test with fact "F" since ?pttrf already tests for these.
```
```suggestion
    # no need to test with fact "F" since ?gttrf already tests for these.
```

mdhaber reviewed (commented):
Looks pretty good.

swallan committed:
TST: refine comments, add lambda missing arguments

mdhaber commented on scipy/linalg/tests/test_lapack.py:
```diff
@@ -1705,3 +1705,201 @@ def test_getc2_gesc2():
+@pytest.mark.parametrize("dtype", DTYPES)
+@pytest.mark.parametrize("fact,dl_d_du_lambda",
+                         [("F", lambda dl, d, du: get_lapack_funcs('gttrf',
+                                                                   dtype=d.dtype)(dl, d, du)),
```
```suggestion
                                                                   dtype=d.dtype)[0](dl, d, du)),
```
Doesn't it return a list?

mdhaber commented on scipy/linalg/tests/test_lapack.py:
```diff
@@ -1705,3 +1705,201 @@ def test_getc2_gesc2():
+    rtol = 250 * np.finfo(dtype).eps
+    atol = 100 * np.finfo(dtype).eps
+    # obtain routine
+    gtsvx = get_lapack_funcs('gtsvx', dtype=dtype)
```
```suggestion
    gtsvx = get_lapack_funcs('gtsvx', dtype=dtype)[0]
```
I think it returns a list.

swallan replied:
It seems to be working for me without that, for me it is returning a fortran object.

I'll have to commit changes tomorrow morning, as I've got to go to my 7 am tomorrow. Sorry that this wasn't fully finished by today.

Kai-Striega replied:
@mdhaber this is fine. `get_lapack_funcs` returns the fortran function directly if `names` is a string:
```python
Python 3.7.4 (default, Aug  9 2019, 18:34:13) [MSC v.1915 64 bit (AMD64)]
Type 'copyright', 'credits' or 'license' for more information
IPython 7.12.0 -- An enhanced Interactive Python. Type '?' for help.

In [1]: import numpy as np
In [2]: from scipy.linalg import get_lapack_funcs
In [3]: get_lapack_funcs('gtsv', dtype=float)
Out[3]: <fortran object>
In [4]: get_lapack_funcs(('gtsv',), dtype=float)
Out[4]: [<fortran object>]
```

mdhaber replied:
Ah.

mdhaber commented on scipy/linalg/tests/test_lapack.py:
```diff
@@ -1705,3 +1705,201 @@ def test_getc2_gesc2():
+    x = generate_random_dtype_array((n, 2), dtype=dtype)
+    # create b from x for equation Ax=b
+    b = A @ x
+    # store a copy of the inputs
```
```suggestion
    # store a copy of the inputs so we can check they haven't been modified later
```

mdhaber commented on scipy/linalg/tests/test_lapack.py:
```diff
@@ -1705,3 +1705,201 @@ def test_getc2_gesc2():
+    inputs_cpy = [dl.copy(), d.copy(), du.copy(), b.copy()]
+
+    # determine dlf, df, duf, du2f, and IPIV from lambda expression
+    dlf_, df_, duf_, du2f_, ipiv_ = dl_d_du_lambda(dl, d, du)
```
?gttrf returns info, right? You need six outputs.

mdhaber commented on scipy/linalg/tests/test_lapack.py:
```diff
@@ -1705,3 +1705,201 @@ def test_getc2_gesc2():
+    # right multiply by final permutation matrix
+    L[:, [i, piv]] = L[:, [piv, i]]
+
+    # check that the outputs define an LU decomposition of A
```
Only needed with `fact='N'`

mdhaber commented on scipy/linalg/tests/test_lapack.py:
```diff
@@ -1705,3 +1705,201 @@ def test_getc2_gesc2():
+            .format(ferr.shape[0], b.shape[1]))
+    # berr should be length of # of cols in x
+    assert_(berr.shape[0] == b.shape[1], "berr.shape is {} but shoud be {},"
+            .format(berr.shape[0], b.shape[1]))
```
Check for `info=0`
Please review this against the original issue gh-11373.

mdhaber commented on scipy/linalg/tests/test_lapack.py:
```diff
@@ -1705,3 +1705,201 @@ def test_getc2_gesc2():
+    # test with singular matrix
+    # no need to test with fact "F" since ?gttrf already tests for these.
+    if fact == "N":
+        d[n-1] = 0
```
```suggestion
        d[-1] = 0
```

mdhaber replied:
I'm glad you agree, but please also implement the change.

mdhaber commented on scipy/linalg/tests/test_lapack.py:
```diff
@@ -1705,3 +1705,201 @@ def test_getc2_gesc2():
+    # no need to test with fact "F" since ?gttrf already tests for these.
+    if fact == "N":
+        d[n-1] = 0
+        dl[n-2] = 0
```
```suggestion
        dl[-2] = 0
```
Below, too.

mdhaber commented on scipy/linalg/tests/test_lapack.py:
```diff
@@ -1705,3 +1705,201 @@ def test_getc2_gesc2():
+
+    else:
+        # assuming that someone is using a singular factorization
+        df[n-1] = 0
```
Should these have underscores? You're modifying `df` and passing `df_` into `gtsvx`. I'm trying to catch things like this, but please trace each of the variables through the code, too.

swallan replied:
Sorry I missed that.

mdhaber commented on scipy/linalg/tests/test_lapack.py:
```diff
@@ -1705,3 +1705,201 @@ def test_getc2_gesc2():
+    A_trans = A.conj().T
+    # Test with transposed matricies
+    # restore these
+    du = inputs_cpy[2]
```
```suggestion
    dl, d, du, _ = inputs_cpy
```

mdhaber commented on scipy/linalg/tests/test_lapack.py:
```diff
@@ -1705,3 +1705,201 @@ def test_getc2_gesc2():
+                "?gtsvx: df[info-1] is {}, not the illegal value"
+                .format(d[info - 1]))
+
+    if dtype in REAL_DTYPES:
```
Can we parameterize the trans options?

swallan replied:
I considered that, but didn't since the complex types and real types take in different trans values, so I thought that there would need to be some sort of if statement anyways. I believe we made the same choice in the test for gttrf/s.

mdhaber commented on scipy/linalg/tests/test_lapack.py:
```diff
@@ -1705,3 +1705,201 @@ def test_getc2_gesc2():
+    # test that x_soln matches the expected x
+    assert_allclose(x, x_soln, rtol=rtol, atol=atol)
+
+    U = np.diag(df, 0) + np.diag(duf, 1) + np.diag(du2f, 2)
```
Since we're checking that the factors represent an LU decomposition in `gttrf`, too, please move all this into a function `L, U = gttrf2lu(dlf, df, duf, du2f, ipiv)` (or similar). Use the function above (and here if you can't parameterize transposal option`, and after the `gttrf` is merged, we should modify its test to use the function.

mdhaber commented on scipy/linalg/tests/test_lapack.py:
```diff
@@ -1705,3 +1705,201 @@ def test_getc2_gesc2():
+    assert_allclose(A, L @ U, atol=atol, rtol=rtol)
+
+
+@pytest.mark.parametrize("du,d,dl,b,x", [(np.array([2.1, -1.0, 1.9, 8.0]),
```
Should probably break the line after `@pytest.mark.parametrize("du,d,dl,b,x",` so there is more space.

swallan replied:
Good call.

mdhaber commented on scipy/linalg/tests/test_lapack.py:
```diff
@@ -1705,3 +1705,201 @@ def test_getc2_gesc2():
+    return np.random.rand(*shape).astype(dtype)
+
+
+@pytest.mark.parametrize("dtype", DTYPES)
```
@swallan you could parameterize the trans values like:

```suggestion
@pytest.mark.parametrize("dtype,trans", 
                                            list(zip(DTYPES, ["N"]*len(DTYPES))) 
                                            + list(zip(REAL_DTYPES , ["T"]*len(REAL_DTYPES ))) 
                                            + list(zip(COMPLEX_DTYPES , ["C"]*len(COMPLEX_DTYPES ))))
```
or something like that.

mdhaber replied:
If you don't want to run the whole test with each transpose option, you could also do something like:
```suggestion
@pytest.mark.parametrize("dtype, trans", ["T", "T", "C", "C"])
```

swallan replied:
The first one seems to work pretty well, I like it.

swallan commented:
@mdhaber Interestingly, when passing in a singular factored matrix (fact=F), it is returning info of n+1 to indicate that RCOND is less than machine precision. Is this particular to my specific machine, or how should I deal with that?

mdhaber commented:
>  Is this particular to my specific machine, or how should I deal with that?

Probably not. The matrix is exactly singular, so something has to go wrong. Check for that in the test.

swallan committed:
TST: Add gttrf lu function, parametrize trans

This commit cleans up small typos, adds a gttrf function to build L and U from factorized diagonals, and was tested with the gttrf lambda.

swallan commented:
> Check for that in the test

OK. 

I have run this on a branch with gttrf merged into it, and it appears to run smoothly, other than the singular matrix returning info of n+1.  I double checked to see that the determinant was zero, so it definitely is exactly singular. I also plan on coming into office hours tomorrow, BTW.

Kai-Striega commented:
@swallan sorry about the delay. I suspect the `n + 1` info is due to (not) converting to/from 1 based indexes used in Fortran.

Kai-Striega commented:
@mdhaber I see that requested some changes, are you satisfied that they have been implemented? And are there any other changes you would like to see or are you happy for me to merge this?

I am also considering converting the `info` argument to be `n` not `n + 1`, however I think that's a simple enough fix for me to make on my branch.

mdhaber commented on scipy/linalg/tests/test_lapack.py:
```diff
@@ -1705,3 +1705,167 @@ def test_getc2_gesc2():
+        dlf, df, duf, du2f, ipiv, x_soln, rcond, ferr, berr, info = gtsvx_out
+        # test for the singular matrix.
+        assert_(d[info - 1] == 0,
+                "?gtsvx: d[info-1] is {}, not the illegal value"
```
@swallan I saw this assert message in the last PR, too. What does it mean?

I think I understand what you're going for. If this assert fails, then `info` is not correct, because `d[info-1]` is not the problem. But I'm not sure that's the right thing to assert. The documentation states that:
```
If some U(i,i)=0, so that U is exactly singular, then the routine
    returns with INFO = i
```
It doesn't say:
```
If some d(i)=0, so that A is exactly singular, then the routine
    returns with INFO = i
```
I'm not certain, but I don't think it's correct in general to assume that because d(i) = 0 that U(i, i) will be zero. The matrix may be permuted before it is factored. Personally, I would just assert that `info > 0`, but if really want to make the test more specific, I think it would be more appropriate to form U and check than U(i,i) is 0.

swallan replied:
> don't think it's correct in general to assume that because d(i) = 0 that U(i, i) will be zero

I see that this might be an issue–I had assumed it and it appeared to be correct for this specific case, but we can just stick with `info > 0` so it would always work.

mdhaber commented on scipy/linalg/tests/test_lapack.py:
```diff
@@ -1705,3 +1705,167 @@ def test_getc2_gesc2():
+    This test uses ?gtsvx to solve a random Ax=b system for each dtype.
+    It tests that the outputs define an LU matrix, that inputs are unmodified,
+    transposal options, incompatible shapes, singular matrices, and
+    singular factorizations. It parametrizes DTYPES and the fact value along
```
```suggestion
    singular factorizations. It parametrizes DTYPES and the `fact` value along
```
I would do something to indicate that "fact" is the name of a variable, not some jargon that the reader doesn't understand.

mdhaber commented on scipy/linalg/tests/test_lapack.py:
```diff
@@ -1705,3 +1705,167 @@ def test_getc2_gesc2():
+
+    # assert that the outputs are of correct type or shape
+    # rcond should be a scalar
+
```
```suggestion
```

mdhaber commented on scipy/linalg/tests/test_lapack.py:
```diff
@@ -1705,3 +1705,167 @@ def test_getc2_gesc2():
+                          du2=du2f_, ipiv=ipiv_)
+        dlf, df, duf, du2f, ipiv, x_soln, rcond, ferr, berr, info = gtsvx_out
+        # info should not be zero and should provide index of illegal value
+        assert_(0 <= info < n, "incorrect info for singular matrix")
```
```suggestion
        assert_(0 < info <= n, "incorrect info for singular matrix")
```
Right? We haven't subtracted 1 from `info` yet.

mdhaber reviewed (changes requested):
See comments. Also, does this test for incorrect/incompatible array sizes for each array? Maybe should split into separate tests to make that more noticeable and to respond to comments in other PRs about breaking down the megatests.

Also, @Kai-Striega did you add `class TestGtsvx`? 
I think it duplicates what is in this PR.

swallan committed:
TST: Break up megatest, array size test

This commit breaks up the test into multiple more specific tests including the test for incompatible array sizes.

swallan commented:
Split, and checked for array sizes. 

@Kai-Striega, Should the error message for incorrectly sized array `b` ValueError? It seemed to raise a `_flapack.error`.

mdhaber commented:
Thanks for these changes. There were also a few places where I suggested `[n-1]` should be `[-1]` and similar.

swallan committed:
TST: change indexing

Kai-Striega commented:
@swallan the ``_flapack.error`` is fine, that occurs when a check in the wrapper isn't true. You can catch these by using the more general ``Exception`` rather than a `ValueError`

swallan committed:
TST: generalize error handling for wrong shape b

Kai-Striega commented:
@mdhaber I did add the tests for gtsvx class. We can leave it in here and I'll remove it once merged.

mdhaber commented:
>We can leave it in here and I'll remove it once merged.

Sure, sounds good.

Kai-Striega commented on scipy/linalg/tests/test_lapack.py:
```diff
@@ -1705,3 +1705,267 @@ def test_getc2_gesc2():
+            .format(berr.shape[0], b.shape[1]))
+
+
+@pytest.mark.parametrize("dtype,trans", list(zip(DTYPES, ["N"]*len(DTYPES)))
+                         + list(zip(REAL_DTYPES, ["T"]*len(REAL_DTYPES)))
+                         + list(zip(COMPLEX_DTYPES,
+                                    ["C"]*len(COMPLEX_DTYPES))))
```
I'm having trouble trying to understand what you are trying to do here. If I'm correct you want to avoid testing the complex conjugates with real data types. In that case it could be easier to use a list comprehension similar to that I used in ?tbtrs:

```python
@pytest.mark.parametrize('dtype,trans',
    [(dtype, trans)
    for dtype in DTYPES for trans in ['N', 'T', 'C']
    if not (trans == 'C' and dtype in REAL_DTYPES)])
```

mdhaber replied:
This code was my suggestion, actually. It tests "N" with all data types, it tests "T" with the real data types, and it tests "C" with the complex data types.  Should we be testing  "T" with all data types?

Kai-Striega commented on scipy/linalg/tests/test_lapack.py:
```diff
@@ -1705,3 +1705,267 @@ def test_getc2_gesc2():
+                                                                        d,
+                                                                        du)),
+                          ("N", lambda dl, d, du: (None, None, None, None,
+                                                   None, None))])
```
I also think this could be neatened up a little. Defining it inside the parameter already uses quite a bit of white space, maybe if you define it as a function, as you did with `gttrf_to_lu` above 👍  will help? OTOH some people may not like having the additional top-level functions. @mdhaber any thoughts?

mdhaber replied:
Yeah that looks really ugly ; ) Let's define it outside @parametrize.

swallan replied:
Ok!

Kai-Striega commented on scipy/linalg/tests/test_lapack.py:
```diff
@@ -1705,3 +1705,267 @@ def test_getc2_gesc2():
+        b = A @ x
+    else:
+        b = (A.conj().T) @ x
+    # determine dlf, df, duf, du2f, and IPIV from lambda expression
```
This comment tells me *what* the line below is doing. I can already see that from the line below. What I really need to understand from the comment is *why* are we doing that?

swallan replied:
The lambda expression was my idea for how to parametrize the `fact` value, so that instead of branching based on` fact == N or T` every time, we could use the lambda to get `dlf_, df_, duf_, du2f_, ipiv_, info_lambda` where when `fact = N`, the lambda returns Nones, and when `fact = T`, lambda returns the output of gttrf.

update: a new function will be used for this instead of parametrization.

mdhaber replied:
@swallan can you respond to @Kai-Striega's suggestion by updating the comment?

mdhaber commented:
@swallan please do implement these when you get a chance so @Kai-Striega can merge.

swallan committed:
TST: remove ugly parametrization, put in function

removed a multiline and ugly parametrization of a lambda, instead going to use a dedicated function outside the test.

swallan committed:
TST: clarify comments on factorization

clarifies comments to show how fact = N or F factorization is done.

mdhaber commented:
I'm going to merge these two in my own branch and submit a PR.
